docs/RFCS/20211207_graceful_draining.md

```diff
@@ -0,0 +1,422 @@
+  - So far no, since we are following the logic of existing `drain_wait` and
+    `query_wait`.
+- Does the change create a new way to communicate data over the network?
+  What rules are in place to ensure that this cannot be used by a malicious
+  user to extract confidential data?
+  - No, it does not create a new communication way.
+- Is there telemetry or crash reporting? What mechanisms are used to ensure
+  no sensitive data is accidentally exposed?
+  - Don’t think it is very related.
+- Is the change visible to users of CockroachDB or operators who run
+  CockroachDB clusters?
+  - Yes, they will notice it from the changes of the reporting system.
+- Is usage of the new feature observable in telemetry? If so, mention where
+  in the code telemetry counters or metrics would be added.
+  - No, it won’t affect telemetry.
+- How to accomodate the use case of changefeed?
+  - Per [Issue 73274][Issue 73274], in the case of changefeed, when the traffic
+    is down to zero (i.e. no ongoing queries), the CDC jobs still needs the server
+    to be alive for a specific period, until it finally signals that it’s “done”
+    and it’s safe to suspend. Such “signal” mechanism is not implemented yet in
+    CDC, but once it’s finished we need to leverage it for the design of
+    `query_wait`.
+
+# Docs changes
+  Since we expect the user to approriately set the length of these three waiting
+  period during the draining process, sufficient tutorial is needed in the
+  documentations.
+
+  Specifically, we should add the following in
+  [Cluster Settings Page][Cluster Setting Page]:
+  - The CockroachDB server's draining procedures. I.e. the behavior of the
+    system during each waiting period. We can refer to the `final state of this
+    design` in this RFC.
+  - How to tune the length of each waiting period when encounter errors in
+    draining.
+      - `server.shutdown.drain_wait` should be longer than the load balancer's
+        health check interval.
+      - `server.shutdown.connection_wait` should be longer than the maximum life
+        time of a connection as specified in the connection pool.
+      - `server.shutdown.query_wait` is more based on user's own preference --
+        longer `query_wait` can give guarantee that on-going queries will be
+        committed before the server fully shutdown, but may lead to a longer
+        draining process. We should also note that `query_wait` let the server
+        exit early once all queries are finished.
+
+# Drawbacks
+- Customer will still receive error messages if their connection pool
+  configuration is not compatible with their wait period settings, e.g.
+  connection pool’s maximum lifetime > `connection_wait`.
+
+# Rationale and Alternative
+The design of a `connection_wait` that allows early exit provides the most
+similar functionality as
+[Postgres’s smart shutdown service][postgres_shutdown].
+
+Note that in the current design, we do support an unlimited `connection_wait`
```

Comment:
what did you end up deciding about enforcing a max allowed value for the new cluster setting? (line 224 mentioned that infinite won't be allowed)